### ISSUE TITLE
Support OpenStack Identity API V3 

### ISSUE THREAD

**Issue by sequoiar** [pkgcloud/pkgcloud#477]

**Comment by stenstad** [comment]
Since the release of Openstack Havana the preferred version of Identity to use is v3. Some cloud providers have now migrated to only supporting v3 for token generation to get support for domains.

Currently pkgcloud does not work with providers that require support for Identity v3.

Would it be possible to add support for Identity v3 in this project as well?

**Comment by BobDickinson** [comment]
IBM BlueMix is one of the providers that stopped supporting v2 recently.  Here is an issue from their support forums:

https://developer.ibm.com/answers/questions/247368/401-unathorised-error-with-object-storage-using-pk.html#answer-256041

There is a guy who hacked together a solution that seems to work (based on mods to pkgcloud - though he did not publish a fork):

http://stackoverflow.com/questions/34415631/bluemix-object-storage-node-js-pkgcloud-openstack-returns-401/34415632#34415632

The package is:

https://www.npmjs.com/package/pkgcloud-bluemix-objectstorage

It's not linked to a source repository, so who knows how it works.  I have tried it, and it does appear to work.

Anyway, not supporting BlueMix storage is a pretty big deal, IMO.  I'm surprised IBM broke the ability of Node.js to talk to BlueMix storage via their (still) recommended approach (using pkgcloud).

If, hypothetically, the working solution from the mystery package could be turned into a PR, is there any chance that would see the light of day in a published release to npm anytime soon?

**Comment by jcrugzz** [comment]
Would love a pull-request for this!

**Comment by BobDickinson** [comment]
I'm working on a PR (I reached out to the guy who created the patched package to see if he wanted to do it, and if not, I'll take a stab at it).

But I'm curious about the chances of that PR solving my problem, given that there seems to be relatively steady maintenance of this repo, but it hasn't been pushed to npm in over a year.  Is there a release plan for 1.2.0?

**Comment by BobDickinson** [comment]
Not sure how I missed this, but this PR from @blueShuttle (6 months ago) looks like the solution:

https://github.com/pkgcloud/pkgcloud/pull/461/files

**Comment by BobDickinson** [comment]
I reviewed the changes in the PR and (while not an expert in openstack), I think it looks good.  It defaults to using v2 auth, so the impact to existing v2 users (everyone) should be minimal. 

I got the PR and ran our storage test suite against BlueMix with it, and it worked great.  

To use v3 auth (required by at least BlueMix) with the PR you just have to add  keystoneAuthVersion: 'v3' as well as your domainId and domainName to your client opts.  

Anyway, it's working great for us.  Sure wish it was merged and pushed to npm ;)
